---
title: "Assembly-based hgcA analysis"
author: "Benjamin D. Peterson"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list = ls())
library(dplyr)
library(knitr)
```


```{r hgcA counts, echo = FALSE}

hgcA.list <- readLines("/Users/benjaminpeterson/Documents/research/Everglades/dataEdited/2018_analysis_assembly/hgcA/hgcA.txt")
hgcA.scaffolds <- paste(hgcA.list %>% strsplit("_", 1) %>% sapply("[", 1),
                        hgcA.list %>% strsplit("_", 1) %>% sapply("[", 2),
                        sep = "_")


hgcB.list <- readLines("/Users/benjaminpeterson/Documents/research/Everglades/dataEdited/2018_analysis_assembly/hgcA/hgcB/hgcB.txt")
hgcB.scaffolds <- paste(hgcB.list %>% strsplit("_", 1) %>% sapply("[", 1),
                        hgcB.list %>% strsplit("_", 1) %>% sapply("[", 2),
                        sep = "_")

hgcA.with.hgcB <- hgcA.scaffolds[which(hgcA.scaffolds %in% hgcB.scaffolds)]

downstream.genes <- readLines("/Users/benjaminpeterson/Documents/research/Everglades/dataEdited/2018_analysis_assembly/hgcA/hgcB/downstream_gene_list.txt")
scaffolds.with.downstream.genes <- paste(downstream.genes %>% strsplit("_", 1) %>% sapply("[", 1),
                                         downstream.genes %>% strsplit("_", 1) %>% sapply("[", 2),
                                         sep = "_")

non.hgcB.downstream.genes <- scaffolds.with.downstream.genes[!(scaffolds.with.downstream.genes %in% hgcB.scaffolds)]

```

After quality-filtering and dereplicating, we obtained `r length(hgcA.list)` hgcA sequences.
Of these, `r length(hgcA.with.hgcB)` have a downstream hgcB sequence.
All of the hgcA sequences without hgcB don't have any ORFs downstream of hgcA, so it's likely that they just didn't assemble into the scaffold, except for one.
One hgcA sequence (`r hgcA.list[hgcA.scaffolds %in% non.hgcB.downstream.genes]`) has a downstream ORF, but it isn't hgcB.
This amino acid (Pw01Meta18_000000010315_3) is only 38 residues long, and on the [Genome Net site](https://www.genome.jp/tools/motif/) it was identified as a cation transport ATPase.
Looking ahead to the phylogeny, it was within the syntroph clade, was very closely related to another hgcA sequence.
It's not an issue with that sequence being cut off either, since the scaffold is 1817bp long and there are three total ORFs downstream of the hgcA sequence, so that one is just missing its hgcB partner.

Alright, time to take a closer look at what these sequences look like.
I built an hgcA phylogeny (below) and attached the abundance information.

```{r hgcA phylogeny and abundance, echo = FALSE}
include_graphics("/Users/benjaminpeterson/Documents/research/Everglades/results/2018_analysis_assembly/hgcA/phylogeny/hgcA_tree_abundance.png")
```

Based on this tree, I assigned a phylogenetic cluster to each sequence.
I generated a summary table of these different groups below.

```{r hgcA taxonomy summary table, echo = FALSE}
include_graphics("/Users/benjaminpeterson/Documents/research/Everglades/results/2018_analysis_assembly/hgcA/depth/taxonomic_cluster_summary_stats.png")

```

As we can see from the tree, the number of sequences and the overall coverage across the 5 metagenomes is dominated by methanogens.
Of the 25 hgcA sequences, 17 are methanogens, and they also account for over two-thirds of the total hgcA coverage.
Next up are the syntrophs, which are represented by 5 hgcA sequences, and account for a little less than a quarter of the coverage.
There are two Geobacter-associated and a single Elusimicrobia-associated hgcA sequences.

Across our sulfate gradient, we see shifts in the guilds that the hgcA sequences are associated with.
Here, I looked at the cumulative abundance of four phylogenetic clusters of hgcA sequences I identified across the five metagenomes.

```{r cumulative hgcA abundance in guilds, echo = FALSE}
include_graphics("/Users/benjaminpeterson/Documents/research/Everglades/results/2018_analysis_assembly/hgcA/depth/depth_by_taxonomic_cluster_plot.png")
```

As we saw in the overall abundances, the *hgcA* sequence coverage is dominated by methanogenic archaea.
However, in the regions with higher levels of sulfate (mostly the 2A transcects), we see much higher levels of Syntroph-associated HgcA sequences.
There is only one Elusimicrobia sequence, so it's difficult to interpret too much from that pattern, but that one is most abundant at intermediate sulfate levels.
Same goes for the Geobacter sequences, although those ones are more highly abundant at the 3A-F site.
The methanogens completely dominate the hgcA coverage at Wagon Wheel, but the overall coverage is super low here.
This site was very sandy, poor quality growth substrate, so might be somewhat oxic deep into the sediments there.

I also wanted to look at the coverage of each individual sequence.
Due to the number of plots, I removed the x-labels, but the sites are in the same order as above.

```{r coverage of individual hgcA sequences, echo = FALSE}
include_graphics("/Users/benjaminpeterson/Documents/research/Everglades/results/2018_analysis_assembly/hgcA/depth/individual_seq_coverage.png")
```

Obviously the Elusimicrobia sequence is gonna be the same since we only had one sequence.
For the Geobacter, both individual sequences matched the overall trend fairly well.
For the methanogens, it might be worth trying to subdivide up the phylogenetic groupings to see if any patterns fall out of that.
As is, it's all over the board.
Some seqs are more abundant at high sulfate concentration, some seem to prefer the lower sulfate concentration sites.
None explicitly prefer Wagon Wheel, but that's to be expected.
Four of the 5 syntroph-associated sequences follow the sulfate gradient closely, most abundant at WCA-2A-P, and decreasing down to 0 at Wagon Wheel.


